### ISSUE TITLE
Method alias annotations are not inferred if the base class has some annotations for them

### ISSUE THREAD

**Issue by ozars** [microsoft/pyright#2385]
**Describe the bug**

Method alias annotations are incorrectly inferred if the alias has some type hints on the base class. This looks like a tricky corner case, not exactly sure if pyright is wrong here as well.

I came across with this problem while using [pytorch](https://github.com/pytorch/pytorch/blob/v1.9.0/torch/nn/modules/module.py#L1101). See below for a simplified example. pytorch relies on a subset of type hints for jit compilation, so static type checking on editor comes quite handful for ensuring jit compilation of complex functions.

**To Reproduce**

```py
from __future__ import annotations
from typing import TYPE_CHECKING, Any, Callable

class Module:
    def _call_impl(self, *args, **kwargs):
        ...

    __call__: Callable[..., Any] = _call_impl
    #  __call__ = _call_impl

class Net(Module):
    def forward(self, x: int) -> int:
        return x + 1

    if TYPE_CHECKING:
        __call__ = forward

model = Net()
out: int | None
out = model(1)

reveal_type(out)
```

```
/home/omer/tmp/pyright/alias.py
  /home/omer/tmp/pyright/alias.py:22:13 - info: Type of "out" is "int | None"
```

**Expected behavior**

Modify above code such that:
```py
    # __call__: Callable[..., Any] = _call_impl
    __call__ = _call_impl
```

```
/home/omer/tmp/pyright/alias.py
  /home/omer/tmp/pyright/alias.py:22:13 - info: Type of "out" is "int"
```

**VS Code extension or command-line**
pyright 1.1.174 CLI

**Comment by erictraut** [comment]
Inference is used only in cases where type annotations are omitted. In this code sample, type annotations are present, and pyright is simply applying those rules. Since the `__call__` symbol in the subclass has no type annotation, the type of `__call__` comes from the annotation in the base class. If you want to provide a different (narrower) type in the subclass, you can add a type annotation in that spot. Either of these would work:

```python
    if TYPE_CHECKING:
        def __call__(self, x: int) -> int: ...
```

or

```python
        __call__: Callable[..., int] = forward
```

**Comment by ozars** [comment]
Makes sense. Thanks. Closing.
